### DISCUSSION TITLE
Interface for random numbers

### DISCUSSION THREAD

**Discussion by jeroenvandijk** [clj-easy/tools.misc#6] [Ideas]
Usage of `(rand)`, `(rand-int n)`, etc are already standardized in Clojure code, but it doesn't give us control over the random seed. Therefore any reliance on random numbers in a library make a program harder to control in a test environment.

If we would have `clj-easy.tools.random` with `rand`, `rand-int` etc and being able to choose the provider we might give this control back to the user. Like with `test.check` we will be able to set a seed and this gives us reproducable/deterministic tests  (and output)

Some remarks
- random function are often used in tight loops so this interface should have minimal overhead (maybe something like swappable libraries as with Polylith and macros to perform things at compile time)
- random number generation is stateful so it will have to be threadsafe (when parallel testing)

**Comment by borkdude** [comment]
Can you name a few existing implementation for which tools.misc should have a facade? Is there something in the JDK that can already be used and why would one want to replace this with another impl?

**Reply by jeroenvandijk** [reply]
The JDK provides [`setSeed`]( https://docs.oracle.com/javase/8/docs/api/java/util/Random.html#setSeed-long-) for instance, but clojure relies on [`Math/random`](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#random--). Then for security tokens it is recommended to use [`SecureRandom`](https://docs.oracle.com/javase/8/docs/api/java/security/SecureRandom.html) or maybe even a custom library. For some of these cases you can set the seed **if** you can get access to the random number generator, if it possible it is probably through monkey patching.

## Some examples of libraries
- [buddy-sign](https://github.com/funcool/buddy-sign/search?q=random) 
- Juxt apex ([nonce generation](https://github.com/juxt/apex/blob/cb4c0016d817ebb03d4f84e572ca71edd84ed79d/modules/oauth2/src/juxt/apex/alpha/oauth2/oic.clj#L60-L68))

## Why
Internally for a long time, we have had a random number generator service in our application. All code relying on random numbers would use this (also for things like uuid generation). This makes tests predictable (always the same coin toss). So you can test on the output and don't have to filter anything. If something changes you know it wasn't because of random numbers. If something fails you can repeat and know you get the same results.

**Reply by borkdude** [reply]
Good example!
